src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let clone = Object.assign({}, state);
+  for (let i = 0; i < actions.length; i++) {
+    const { type, extraData, keysToRemove } = actions[i];
+
+    if (i > 0) {
+      clone = Object.assign({}, newState[i - 1]);
+    }
+
+
+    if (type === 'addProperties') {
+      newState.push({ ...clone, ...extraData });
+    }
```

Comment:
[CODE STYLE] - Rather than using an if statement for each type, use a switch statement to handle different action types. This will make the code cleaner and more maintainable. Ensure you have a default case for error handling. ([CODE STYLE]: Use switch statement if you have limited amount of conditions. [CODE STYLE]: switch/case should always have default case for error handling.)